src/canary.rs

```diff
@@ -33,76 +33,127 @@ const CANARY_VALUE: u8 = 0xAA;
         core.reset_and_halt(TIMEOUT)?;
 
         // Decide if and where to place the stack canary.
-        if let Some(highest_static_var_address) = target_info.highest_static_var_address {
-            // standard = static variables are at a lower address; stack (grows down) is at a higher address
-            let standard_memory_layout =
-                highest_static_var_address < elf.vector_table.initial_stack_pointer;
-
-            if !elf.program_uses_heap() && standard_memory_layout {
-                let stack_available =
-                    elf.vector_table.initial_stack_pointer - highest_static_var_address - 1;
-
-                // We consider >90% stack usage a potential stack overflow, but don't go beyond 1 kb since
-                // filling a lot of RAM is slow (and 1 kb should be "good enough" for what we're doing).
-                let size = 1024.min(stack_available / 10) as usize;
-
-                log::debug!(
-                    "{} bytes of stack available ({:#010X} ..= {:#010X}), using {} byte canary to detect overflows",
-                    stack_available,
-                    highest_static_var_address + 1,
-                    elf.vector_table.initial_stack_pointer,
-                    size,
-                );
-
-                // Canary starts right after `highest_ram_addr_in_use`.
-                let address = highest_static_var_address + 1;
-                let canary = vec![CANARY_VALUE; size];
-                core.write_8(address, &canary)?;
-
-                return Ok(Some(Canary { address, size }));
+
+        let stack_range = match &target_info.stack_range {
+            Some(range) => range,
+            None => {
+                log::debug!("couldn't find valid stack range, not placing stack canary");
+                return Ok(None);
             }
+        };
+
+        if elf.program_uses_heap() {
+            log::debug!("heap in use, not placing stack canary");
+            return Ok(None);
+        }
+
+        let stack_available = stack_range.end() - stack_range.start() - 1;
+
+        let size = if measure_stack {
+            // When measuring stack consumption, we have to color the whole stack.
+            stack_available as usize
+        } else {
+            // We consider >90% stack usage a potential stack overflow, but don't go beyond 1 kb
+            // since filling a lot of RAM is slow (and 1 kb should be "good enough" for what we're
+            // doing).
+            1024.min(stack_available / 10) as usize
+        };
+
+        log::debug!(
+            "{} bytes of stack available ({:#010X} ..= {:#010X}), using {} byte canary",
+            stack_available,
+            stack_range.start(),
+            stack_range.end(),
+            size,
+        );
+
+        if measure_stack {
+            // Painting 100KB or more takes a few seconds, so provide user feedback.
+            log::info!("painting {} bytes of RAM for stack usage estimation", size);
```

Comment:
nit: could we report the number of bytes here in KB? either than use a comma to separate the thousands (e.g. `123,456`; see num-format crate). It's a bit hard to read at the moment will all the digits next to each other.